designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,224 @@
+# Disruption Controls By Action 
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. High-Frequency Trading (HFT) firms require full compute capacity during specific operational hours, making it imperative to avoid scale-down requests for consolidation during these periods. However, outside these hours, scale-downs are acceptable.
+
+## Known Requirements and Desired Behaviors 
+**Action and Budget Definition:** Users should be able to define an action and a corresponding budget(s).
+**Supported Actions:** All disruption actions affected by the current Budgets implementation (Consolidation, Emptiness, Expiration, Drift) should be supported.
+**Default Behavior for Unspecified Actions:** Budgets should continue to support a default behavior for all disruption actions. If an action is unspecified, it is assumed to apply to all actions.
+
+### Q: How should Karpenterr handle the default or undefined action case? 
+The current design involves specifying a specific number of disruptable nodes per action, which can complicate the disruption lifecycle. For example, if there's a 10-node budget for "Drift" and a separate 10-node budget for "Consolidation," but a 15-node budget for "All" determining which nodes will get disrupted becomes unclear. Would it be 10 nodes for "Drift" and 5 nodes for "Consolidation"?
+
+#### Unspecified Action as a Global Node Budget
+
+We could consider treating an undefined action as a budget for all disruption actions except for those with explicitly defined budgets. In this scenario, if a user specifies a disruption budget like this:
+
+```yaml
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      action: Consolidation
+      nodes: 10
+    - schedule: "* * * * *"
+      action: Drift
+      nodes: 10
+    # For all other actions , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+```
+
+It means that "Consolidation" and "Drift" actions have specific budgets of 10 nodes each, while all other actions (e.g., expiration and emptiness) share a common budget of 5 nodes. This approach simplifies the configuration but has one limitation: it may not allow the execution of other disruption actions if a specific action exhausts the budget. This is a problem with the existing design for disruption budgets.
+
+There are two ways for the users to get around this behavior. 
+1. If you need gaurenteed disruption for a particular action, you can just specify that action in a budget.  
+2. We could allow some mechanism for the users to control the ordering of the disruption actions. 
+#### Q: Should users be able to change the order that disruption actions are executed in to solve this problem? 
```

Comment:
This is jumping the shark a bit and more of a question for @njtran. With this is there any consideration to allow for disruption actions to happen in parallel? Since each disruption action has a bucket now in theory there is no need for the order (maybe) since in a way we are "ordering" consolidation by giving them individual buckets